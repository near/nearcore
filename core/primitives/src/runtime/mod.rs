pub use near_primitives_core::runtime::fees;
pub use near_primitives_core::runtime::*;

pub mod apply_state;
pub mod config;
pub mod config_store;
pub mod migration_data;
pub mod parameter_table;
