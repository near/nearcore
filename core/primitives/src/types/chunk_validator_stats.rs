use {
    super::{NumBlocks, ValidatorStats},
    borsh::{self, BorshDeserialize, BorshSerialize},
};

/// An extension to `ValidatorStats` which also tracks endorsements
/// coming from stateless validators.
#[derive(Default, BorshSerialize, BorshDeserialize, Clone, Debug, PartialEq, Eq)]
pub struct ChunkStats {
    pub production: ValidatorStats,
    pub endorsement: ValidatorStats,
}

impl ChunkStats {
    pub const fn new(
        chunks_produced: u64,
        chunks_expected: u64,
        endorsements_produced: u64,
        endorsements_expected: u64,
    ) -> Self {
        ChunkStats {
            production: ValidatorStats { produced: chunks_produced, expected: chunks_expected },
            endorsement: ValidatorStats {
                produced: endorsements_produced,
                expected: endorsements_expected,
            },
        }
    }

    pub const fn new_with_production(produced: u64, expected: u64) -> Self {
        ChunkStats {
            production: ValidatorStats { produced, expected },
            endorsement: ValidatorStats { produced: 0, expected: 0 },
        }
    }

    pub const fn new_with_endorsement(produced: u64, expected: u64) -> Self {
        ChunkStats {
            production: ValidatorStats { produced: 0, expected: 0 },
            endorsement: ValidatorStats { produced, expected },
        }
    }

    pub fn produced(&self) -> NumBlocks {
        self.production.produced
    }

    pub fn expected(&self) -> NumBlocks {
        self.production.expected
    }

    pub fn produced_mut(&mut self) -> &mut NumBlocks {
        &mut self.production.produced
    }

    pub fn expected_mut(&mut self) -> &mut NumBlocks {
        &mut self.production.expected
    }

    pub fn production_stats(&self) -> &ValidatorStats {
        &self.production
    }

    pub fn endorsement_stats(&self) -> &ValidatorStats {
        &self.endorsement
    }

    pub fn endorsement_stats_mut(&mut self) -> &mut ValidatorStats {
        &mut self.endorsement
    }
}

#[test]
fn test_mutability() {
    let mut stats = ChunkStats::new_with_production(0, 0);

    *stats.expected_mut() += 1;
    assert_eq!(stats, ChunkStats::new_with_production(0, 1));

    *stats.produced_mut() += 1;
    assert_eq!(stats, ChunkStats::new_with_production(1, 1));

    let endorsement_stats = stats.endorsement_stats_mut();
    endorsement_stats.produced += 10;
    endorsement_stats.expected += 10;

    assert_eq!(
        stats,
        ChunkStats {
            production: ValidatorStats { produced: 1, expected: 1 },
            endorsement: ValidatorStats { produced: 10, expected: 10 }
        }
    );

    *stats.expected_mut() += 1;
    assert_eq!(
        stats,
        ChunkStats {
            production: ValidatorStats { produced: 1, expected: 2 },
            endorsement: ValidatorStats { produced: 10, expected: 10 }
        }
    );

    *stats.produced_mut() += 1;
    assert_eq!(
        stats,
        ChunkStats {
            production: ValidatorStats { produced: 2, expected: 2 },
            endorsement: ValidatorStats { produced: 10, expected: 10 }
        }
    );

    let endorsement_stats = stats.endorsement_stats_mut();
    endorsement_stats.produced += 10;
    endorsement_stats.expected += 10;

    assert_eq!(
        stats,
        ChunkStats {
            production: ValidatorStats { produced: 2, expected: 2 },
            endorsement: ValidatorStats { produced: 20, expected: 20 }
        }
    );
}
