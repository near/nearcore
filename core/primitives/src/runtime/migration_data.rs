use crate::receipt::ReceiptResult;
use crate::types::AccountId;
use crate::types::Gas;
use std::fmt;
use std::fmt::{Debug, Formatter};

#[derive(Default)]
pub struct MigrationData {
    pub storage_usage_delta: Vec<(AccountId, u64)>,
    pub storage_usage_fix_gas: Gas,
    pub restored_receipts: ReceiptResult,
}

impl Debug for MigrationData {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.debug_struct("MigrationData").finish()
    }
}

#[derive(Debug, Default)]
pub struct MigrationFlags {
    // True iff the current block is the first one in the chain with current protocol version
    pub is_first_block_of_version: bool,
    // True iff, among all blocks containing chunk for some specific shard, the current block is the
    // first one in the first epoch with the current protocol version
    pub is_first_block_with_chunk_of_version: bool,
}
