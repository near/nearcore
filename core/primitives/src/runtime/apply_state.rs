use crate::runtime::migration_data::{MigrationData, MigrationFlags};
use crate::{
    hash::CryptoHash,
    runtime::config::RuntimeConfig,
    types::{Balance, BlockHeight, EpochHeight, EpochId, Gas},
    version::ProtocolVersion,
};
use near_vm_runner::logic::CompiledContractCache;
use std::sync::Arc;

#[derive(Debug)]
pub struct ApplyState {
    /// Currently building block height.
    pub block_height: BlockHeight,
    /// Prev block hash
    pub prev_block_hash: CryptoHash,
    /// Current block hash
    pub block_hash: CryptoHash,
    /// Current epoch id
    pub epoch_id: EpochId,
    /// Current epoch height
    pub epoch_height: EpochHeight,
    /// Price for the gas.
    pub gas_price: Balance,
    /// The current block timestamp (number of non-leap-nanoseconds since January 1, 1970 0:00:00 UTC).
    pub block_timestamp: u64,
    /// Gas limit for a given chunk.
    /// If None is given, assumes there is no gas limit.
    pub gas_limit: Option<Gas>,
    /// Current random seed (from current block vrf output).
    pub random_seed: CryptoHash,
    /// Current Protocol version when we apply the state transition
    pub current_protocol_version: ProtocolVersion,
    /// The Runtime config to use for the current transition.
    pub config: Arc<RuntimeConfig>,
    /// Cache for compiled contracts.
    pub cache: Option<Box<dyn CompiledContractCache>>,
    /// Whether the chunk being applied is new.
    pub is_new_chunk: bool,
    /// Data for migrations that may need to be applied at the start of an epoch when protocol
    /// version changes
    pub migration_data: Arc<MigrationData>,
    /// Flags for migrations indicating whether they can be applied at this block
    pub migration_flags: MigrationFlags,
}
