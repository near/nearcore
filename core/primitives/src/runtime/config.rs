//! Settings of the parameters of the runtime.
use crate::config::VMConfig;
use crate::runtime::config_store::INITIAL_TESTNET_CONFIG;
use crate::runtime::fees::RuntimeFeesConfig;
use crate::runtime::parameter_table::ParameterTable;
use crate::types::AccountId;
use near_primitives_core::types::Balance;

use super::parameter_table::InvalidConfigError;

/// The structure that holds the parameters of the runtime, mostly economics.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RuntimeConfig {
    /// Action gas costs, storage fees, and economic constants around them.
    ///
    /// This contains parameters that are required by the WASM runtime and the
    /// transaction runtime.
    pub fees: RuntimeFeesConfig,
    /// Config of wasm operations, also includes wasm gas costs.
    ///
    /// This contains all the configuration parameters that are only required by
    /// the WASM runtime.
    pub wasm_config: VMConfig,
    /// Config that defines rules for account creation.
    pub account_creation_config: AccountCreationConfig,
}

impl RuntimeConfig {
    pub(crate) fn new(params: &ParameterTable) -> Result<Self, InvalidConfigError> {
        RuntimeConfig::try_from(params)
    }

    pub fn initial_testnet_config() -> RuntimeConfig {
        INITIAL_TESTNET_CONFIG
            .parse()
            .and_then(|params| RuntimeConfig::new(&params))
            .expect("Failed parsing initial testnet config")
    }

    pub fn test() -> Self {
        RuntimeConfig {
            fees: RuntimeFeesConfig::test(),
            wasm_config: VMConfig::test(),
            account_creation_config: AccountCreationConfig::default(),
        }
    }

    pub fn free() -> Self {
        Self {
            fees: RuntimeFeesConfig::free(),
            wasm_config: VMConfig::free(),
            account_creation_config: AccountCreationConfig::default(),
        }
    }

    pub fn storage_amount_per_byte(&self) -> Balance {
        self.fees.storage_usage_config.storage_amount_per_byte
    }
}

/// The structure describes configuration for creation of new accounts.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AccountCreationConfig {
    /// The minimum length of the top-level account ID that is allowed to be created by any account.
    pub min_allowed_top_level_account_length: u8,
    /// The account ID of the account registrar. This account ID allowed to create top-level
    /// accounts of any valid length.
    pub registrar_account_id: AccountId,
}

impl Default for AccountCreationConfig {
    fn default() -> Self {
        Self {
            min_allowed_top_level_account_length: 0,
            registrar_account_id: "registrar".parse().unwrap(),
        }
    }
}
