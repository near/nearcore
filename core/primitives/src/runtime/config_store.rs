use crate::runtime::config::RuntimeConfig;
use crate::runtime::parameter_table::{ParameterTable, ParameterTableDiff};
use crate::types::ProtocolVersion;
use std::collections::BTreeMap;
use std::ops::Bound;
use std::sync::Arc;

macro_rules! include_config {
    ($file:expr) => {
        include_str!(concat!("../../res/runtime_configs/", $file))
    };
}

/// The base config file with all initial parameter values defined.
/// Later version are calculated by applying diffs to this base.
static BASE_CONFIG: &str = include_config!("parameters.yaml");

/// Stores pairs of protocol versions for which runtime config was updated and
/// the file containing the diffs in bytes.
static CONFIG_DIFFS: &[(ProtocolVersion, &str)] = &[
    (42, include_config!("42.yaml")),
    (48, include_config!("48.yaml")),
    (49, include_config!("49.yaml")),
    (50, include_config!("50.yaml")),
    // max_gas_burnt increased to 300 TGas
    (52, include_config!("52.yaml")),
    // Increased deployment costs, increased wasmer2 stack_limit, added limiting of contract locals,
    // set read_cached_trie_node cost, decrease storage key limit
    (53, include_config!("53.yaml")),
    (57, include_config!("57.yaml")),
    // Introduce Zero Balance Account and increase account creation cost to 7.7Tgas
    (59, include_config!("59.yaml")),
    (61, include_config!("61.yaml")),
    (62, include_config!("62.yaml")),
];

/// Testnet parameters for versions <= 29, which (incorrectly) differed from mainnet parameters
pub static INITIAL_TESTNET_CONFIG: &str = include_config!("parameters_testnet.yaml");

/// Stores runtime config for each protocol version where it was updated.
#[derive(Debug)]
pub struct RuntimeConfigStore {
    /// Maps protocol version to the config.
    store: BTreeMap<ProtocolVersion, Arc<RuntimeConfig>>,
}

impl RuntimeConfigStore {
    /// Constructs a new store.
    ///
    /// If genesis_runtime_config is Some, configs for protocol versions 0 and 42 are overridden by
    /// this config and config with lowered storage cost, respectively.
    /// This is done to preserve compatibility with previous implementation, where we updated
    /// runtime config by sequential modifications to the genesis runtime config.
    /// calimero_zero_storage flag sets all storages fees to zero by setting
    /// storage_amount_per_byte to zero, to keep calimero private shards compatible with future
    /// protocol upgrades this is done for all protocol versions
    /// TODO #4775: introduce new protocol version to have the same runtime config for all chains
    pub fn new(genesis_runtime_config: Option<&RuntimeConfig>) -> Self {
        let mut params: ParameterTable =
            BASE_CONFIG.parse().expect("Failed parsing base parameter file.");

        let mut store = BTreeMap::new();
        #[cfg(not(feature = "calimero_zero_storage"))]
        {
            let initial_config = RuntimeConfig::new(&params).unwrap_or_else(|err| panic!("Failed generating `RuntimeConfig` from parameters for base parameter file. Error: {err}"));
            store.insert(0, Arc::new(initial_config));
        }
        #[cfg(feature = "calimero_zero_storage")]
        {
            let mut initial_config = RuntimeConfig::new(&params).unwrap_or_else(|err| panic!("Failed generating `RuntimeConfig` from parameters for base parameter file. Error: {err}"));
            initial_config.fees.storage_usage_config.storage_amount_per_byte = 0;
            store.insert(0, Arc::new(initial_config));
        }

        for (protocol_version, diff_bytes) in CONFIG_DIFFS {
            let diff :ParameterTableDiff= diff_bytes.parse().unwrap_or_else(|err| panic!("Failed parsing runtime parameters diff for version {protocol_version}. Error: {err}"));
            params.apply_diff(diff).unwrap_or_else(|err| panic!("Failed applying diff to `RuntimeConfig` for version {protocol_version}. Error: {err}"));
            #[cfg(not(feature = "calimero_zero_storage"))]
            store.insert(
                *protocol_version,
                Arc::new(RuntimeConfig::new(&params).unwrap_or_else(|err| panic!("Failed generating `RuntimeConfig` from parameters for version {protocol_version}. Error: {err}"))),
            );
            #[cfg(feature = "calimero_zero_storage")]
            {
                let mut runtime_config = RuntimeConfig::new(&params).unwrap_or_else(|err| panic!("Failed generating `RuntimeConfig` from parameters for version {protocol_version}. Error: {err}"));
                runtime_config.fees.storage_usage_config.storage_amount_per_byte = 0;
                store.insert(*protocol_version, Arc::new(runtime_config));
            }
        }

        if let Some(runtime_config) = genesis_runtime_config {
            let mut config = runtime_config.clone();
            store.insert(0, Arc::new(config.clone()));

            config.fees.storage_usage_config.storage_amount_per_byte = 10u128.pow(19);
            store.insert(42, Arc::new(config));
        }

        Self { store }
    }

    /// Create store of runtime configs for the given chain id.
    ///
    /// For mainnet and other chains except testnet we don't need to override runtime config for
    /// first protocol versions.
    /// For testnet, runtime config for genesis block was (incorrectly) different, that's why we
    /// need to override it specifically to preserve compatibility.
    pub fn for_chain_id(chain_id: &str) -> Self {
        match chain_id {
            "testnet" => {
                let genesis_runtime_config = RuntimeConfig::initial_testnet_config();
                Self::new(Some(&genesis_runtime_config))
            }
            _ => Self::new(None),
        }
    }

    /// Constructs test store.
    pub fn with_one_config(runtime_config: RuntimeConfig) -> Self {
        Self { store: BTreeMap::from_iter([(0, Arc::new(runtime_config))].iter().cloned()) }
    }

    /// Constructs test store.
    pub fn test() -> Self {
        Self::with_one_config(RuntimeConfig::test())
    }

    /// Constructs store with a single config with zero costs.
    pub fn free() -> Self {
        Self::with_one_config(RuntimeConfig::free())
    }

    /// Returns a `RuntimeConfig` for the corresponding protocol version.
    pub fn get_config(&self, protocol_version: ProtocolVersion) -> &Arc<RuntimeConfig> {
        self.store
            .range((Bound::Unbounded, Bound::Included(protocol_version)))
            .next_back()
            .unwrap_or_else(|| {
                panic!("Not found RuntimeConfig for protocol version {}", protocol_version)
            })
            .1
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::version::ProtocolFeature::{
        LowerDataReceiptAndEcrecoverBaseCost, LowerStorageCost, LowerStorageKeyLimit,
    };
    use near_primitives_core::config::{ActionCosts, ExtCosts};

    const GENESIS_PROTOCOL_VERSION: ProtocolVersion = 29;
    const RECEIPTS_DEPTH: u64 = 63;

    #[test]
    fn test_max_prepaid_gas() {
        let store = RuntimeConfigStore::new(None);
        for (protocol_version, config) in store.store.iter() {
            assert!(
                config.wasm_config.limit_config.max_total_prepaid_gas
                    / config.fees.min_receipt_with_function_call_gas()
                    <= 63,
                "The maximum desired depth of receipts for protocol version {} should be at most {}",
                protocol_version,
                RECEIPTS_DEPTH
            );
        }
    }

    #[test]
    #[cfg(not(feature = "calimero_zero_storage"))]
    fn test_lower_storage_cost() {
        let store = RuntimeConfigStore::new(None);
        let base_cfg = store.get_config(GENESIS_PROTOCOL_VERSION);
        let new_cfg = store.get_config(LowerStorageCost.protocol_version());
        assert!(base_cfg.storage_amount_per_byte() > new_cfg.storage_amount_per_byte());
    }

    #[test]
    fn test_override_account_length() {
        // Check that default value is 32.
        let base_store = RuntimeConfigStore::new(None);
        let base_cfg = base_store.get_config(GENESIS_PROTOCOL_VERSION);
        assert_eq!(base_cfg.account_creation_config.min_allowed_top_level_account_length, 32);

        let mut cfg = base_cfg.as_ref().clone();
        cfg.account_creation_config.min_allowed_top_level_account_length = 0;

        // Check that length was changed.
        let new_store = RuntimeConfigStore::new(Some(&cfg));
        let new_cfg = new_store.get_config(GENESIS_PROTOCOL_VERSION);
        assert_eq!(new_cfg.account_creation_config.min_allowed_top_level_account_length, 0);
    }

    #[test]
    fn test_lower_data_receipt_cost() {
        let store = RuntimeConfigStore::new(None);
        let base_cfg = store.get_config(LowerStorageCost.protocol_version());
        let new_cfg = store.get_config(LowerDataReceiptAndEcrecoverBaseCost.protocol_version());
        assert!(
            base_cfg.fees.fee(ActionCosts::new_data_receipt_base).send_sir
                > new_cfg.fees.fee(ActionCosts::new_data_receipt_base).send_sir
        );
        assert!(
            base_cfg.fees.fee(ActionCosts::new_data_receipt_byte).send_sir
                > new_cfg.fees.fee(ActionCosts::new_data_receipt_byte).send_sir
        );
    }

    // Check that for protocol version with lowered data receipt cost, runtime config passed to
    // config store is overridden.
    #[test]
    #[cfg(not(feature = "calimero_zero_storage"))]
    fn test_override_runtime_config() {
        let store = RuntimeConfigStore::new(None);
        let config = store.get_config(0);

        let mut base_params = BASE_CONFIG.parse().unwrap();
        let base_config = RuntimeConfig::new(&base_params).unwrap();
        assert_eq!(config.as_ref(), &base_config);

        let config = store.get_config(LowerStorageCost.protocol_version());
        assert_eq!(base_config.storage_amount_per_byte(), 100_000_000_000_000_000_000u128);
        assert_eq!(config.storage_amount_per_byte(), 10_000_000_000_000_000_000u128);
        assert_eq!(config.fees.fee(ActionCosts::new_data_receipt_base).send_sir, 4_697_339_419_375);
        assert_ne!(config.as_ref(), &base_config);
        assert_ne!(
            config.as_ref(),
            store.get_config(LowerStorageCost.protocol_version() - 1).as_ref()
        );

        let expected_config = {
            let first_diff = CONFIG_DIFFS[0].1.parse().unwrap();
            base_params.apply_diff(first_diff).unwrap();
            RuntimeConfig::new(&base_params).unwrap()
        };
        assert_eq!(**config, expected_config);

        let config = store.get_config(LowerDataReceiptAndEcrecoverBaseCost.protocol_version());
        assert_eq!(config.fees.fee(ActionCosts::new_data_receipt_base).send_sir, 36_486_732_312);
        let expected_config = {
            let second_diff = CONFIG_DIFFS[1].1.parse().unwrap();
            base_params.apply_diff(second_diff).unwrap();
            RuntimeConfig::new(&base_params).unwrap()
        };
        assert_eq!(config.as_ref(), &expected_config);
    }

    #[test]
    fn test_lower_ecrecover_base_cost() {
        let store = RuntimeConfigStore::new(None);
        let base_cfg = store.get_config(LowerStorageCost.protocol_version());
        let new_cfg = store.get_config(LowerDataReceiptAndEcrecoverBaseCost.protocol_version());
        assert!(
            base_cfg.wasm_config.ext_costs.gas_cost(ExtCosts::ecrecover_base)
                > new_cfg.wasm_config.ext_costs.gas_cost(ExtCosts::ecrecover_base)
        );
    }

    #[test]
    fn test_lower_max_length_storage_key() {
        let store = RuntimeConfigStore::new(None);
        let base_cfg = store.get_config(LowerStorageKeyLimit.protocol_version() - 1);
        let new_cfg = store.get_config(LowerStorageKeyLimit.protocol_version());
        assert!(
            base_cfg.wasm_config.limit_config.max_length_storage_key
                > new_cfg.wasm_config.limit_config.max_length_storage_key
        );
    }

    /// Use snapshot testing to check that the JSON representation of the
    /// configurations of each version is unchanged.
    /// If tests fail after an intended change, run `cargo insta review` accept
    /// the new snapshot if it looks right.
    #[test]
    #[cfg(not(feature = "nightly"))]
    #[cfg(not(feature = "calimero_zero_storage"))]
    fn test_json_unchanged() {
        use crate::views::RuntimeConfigView;

        let store = RuntimeConfigStore::new(None);

        for version in store.store.keys() {
            let snapshot_name = format!("{version}.json");
            let config_view = RuntimeConfigView::from(store.get_config(*version).as_ref().clone());
            insta::assert_json_snapshot!(snapshot_name, config_view);
        }

        // Store the latest values of parameters in a human-readable snapshot.
        {
            let mut params: ParameterTable = BASE_CONFIG.parse().unwrap();
            for (_, diff_bytes) in CONFIG_DIFFS {
                params.apply_diff(diff_bytes.parse().unwrap()).unwrap();
            }
            insta::with_settings!({
                snapshot_path => "../../res/runtime_configs",
                prepend_module_to_snapshot => false,
                description => "THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.",
                omit_expression => true,
            }, {
                insta::assert_display_snapshot!("parameters", params);
            });
        }

        // Testnet initial config for old version was different, thus needs separate testing
        let params = INITIAL_TESTNET_CONFIG.parse().unwrap();
        let new_genesis_runtime_config = RuntimeConfig::new(&params).unwrap();
        let testnet_store = RuntimeConfigStore::new(Some(&new_genesis_runtime_config));

        for version in testnet_store.store.keys() {
            let snapshot_name = format!("testnet_{version}.json");
            let config_view = RuntimeConfigView::from(store.get_config(*version).as_ref().clone());
            insta::assert_json_snapshot!(snapshot_name, config_view);
        }
    }

    #[test]
    #[cfg(feature = "calimero_zero_storage")]
    fn test_calimero_storage_costs_zero() {
        let store = RuntimeConfigStore::new(None);
        for (_, config) in store.store.iter() {
            assert_eq!(config.storage_amount_per_byte(), 0u128);
        }
    }
}
