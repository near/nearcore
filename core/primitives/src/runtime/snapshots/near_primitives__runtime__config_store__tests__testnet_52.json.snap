---
source: core/primitives/src/runtime/config_store.rs
expression: config_view
---
{
  "storage_amount_per_byte": "10000000000000000000",
  "transaction_costs": {
    "action_receipt_creation_config": {
      "send_sir": 108059500000,
      "send_not_sir": 108059500000,
      "execution": 108059500000
    },
    "data_receipt_creation_config": {
      "base_cost": {
        "send_sir": 36486732312,
        "send_not_sir": 36486732312,
        "execution": 36486732312
      },
      "cost_per_byte": {
        "send_sir": 17212011,
        "send_not_sir": 17212011,
        "execution": 17212011
      }
    },
    "action_creation_config": {
      "create_account_cost": {
        "send_sir": 99607375000,
        "send_not_sir": 99607375000,
        "execution": 99607375000
      },
      "deploy_contract_cost": {
        "send_sir": 184765750000,
        "send_not_sir": 184765750000,
        "execution": 184765750000
      },
      "deploy_contract_cost_per_byte": {
        "send_sir": 6812999,
        "send_not_sir": 6812999,
        "execution": 6812999
      },
      "function_call_cost": {
        "send_sir": 2319861500000,
        "send_not_sir": 2319861500000,
        "execution": 2319861500000
      },
      "function_call_cost_per_byte": {
        "send_sir": 2235934,
        "send_not_sir": 2235934,
        "execution": 2235934
      },
      "transfer_cost": {
        "send_sir": 115123062500,
        "send_not_sir": 115123062500,
        "execution": 115123062500
      },
      "stake_cost": {
        "send_sir": 141715687500,
        "send_not_sir": 141715687500,
        "execution": 102217625000
      },
      "add_key_cost": {
        "full_access_cost": {
          "send_sir": 101765125000,
          "send_not_sir": 101765125000,
          "execution": 101765125000
        },
        "function_call_cost": {
          "send_sir": 102217625000,
          "send_not_sir": 102217625000,
          "execution": 102217625000
        },
        "function_call_cost_per_byte": {
          "send_sir": 1925331,
          "send_not_sir": 1925331,
          "execution": 1925331
        }
      },
      "delete_key_cost": {
        "send_sir": 94946625000,
        "send_not_sir": 94946625000,
        "execution": 94946625000
      },
      "delete_account_cost": {
        "send_sir": 147489000000,
        "send_not_sir": 147489000000,
        "execution": 147489000000
      },
      "delegate_cost": {
        "send_sir": 200000000000,
        "send_not_sir": 200000000000,
        "execution": 200000000000
      }
    },
    "storage_usage_config": {
      "num_bytes_account": 100,
      "num_extra_bytes_record": 40
    },
    "burnt_gas_reward": [
      3,
      10
    ],
    "pessimistic_gas_price_inflation_ratio": [
      103,
      100
    ]
  },
  "wasm_config": {
    "ext_costs": {
      "base": 264768111,
      "contract_loading_base": 35445963,
      "contract_loading_bytes": 216750,
      "read_memory_base": 2609863200,
      "read_memory_byte": 3801333,
      "write_memory_base": 2803794861,
      "write_memory_byte": 2723772,
      "read_register_base": 2517165186,
      "read_register_byte": 98562,
      "write_register_base": 2865522486,
      "write_register_byte": 3801564,
      "utf8_decoding_base": 3111779061,
      "utf8_decoding_byte": 291580479,
      "utf16_decoding_base": 3543313050,
      "utf16_decoding_byte": 163577493,
      "sha256_base": 4540970250,
      "sha256_byte": 24117351,
      "keccak256_base": 5879491275,
      "keccak256_byte": 21471105,
      "keccak512_base": 5811388236,
      "keccak512_byte": 36649701,
      "ripemd160_base": 853675086,
      "ripemd160_block": 680107584,
      "ed25519_verify_base": 210000000000,
      "ed25519_verify_byte": 9000000,
      "ecrecover_base": 278821988457,
      "log_base": 3543313050,
      "log_byte": 13198791,
      "storage_write_base": 64196736000,
      "storage_write_key_byte": 70482867,
      "storage_write_value_byte": 31018539,
      "storage_write_evicted_byte": 32117307,
      "storage_read_base": 56356845750,
      "storage_read_key_byte": 30952533,
      "storage_read_value_byte": 5611005,
      "storage_remove_base": 53473030500,
      "storage_remove_key_byte": 38220384,
      "storage_remove_ret_value_byte": 11531556,
      "storage_has_key_base": 54039896625,
      "storage_has_key_byte": 30790845,
      "storage_iter_create_prefix_base": 0,
      "storage_iter_create_prefix_byte": 0,
      "storage_iter_create_range_base": 0,
      "storage_iter_create_from_byte": 0,
      "storage_iter_create_to_byte": 0,
      "storage_iter_next_base": 0,
      "storage_iter_next_key_byte": 0,
      "storage_iter_next_value_byte": 0,
      "touching_trie_node": 16101955926,
      "read_cached_trie_node": 2280000000,
      "promise_and_base": 1465013400,
      "promise_and_per_promise": 5452176,
      "promise_return": 560152386,
      "validator_stake_base": 911834726400,
      "validator_total_stake_base": 911834726400,
      "contract_compile_base": 0,
      "contract_compile_bytes": 0,
      "alt_bn128_g1_multiexp_base": 713000000000,
      "alt_bn128_g1_multiexp_element": 320000000000,
      "alt_bn128_g1_sum_base": 3000000000,
      "alt_bn128_g1_sum_element": 5000000000,
      "alt_bn128_pairing_check_base": 9686000000000,
      "alt_bn128_pairing_check_element": 5102000000000
    },
    "grow_mem_cost": 1,
    "regular_op_cost": 822756,
    "limit_config": {
      "max_gas_burnt": 300000000000000,
      "max_stack_height": 16384,
      "contract_prepare_version": 1,
      "initial_memory_pages": 1024,
      "max_memory_pages": 2048,
      "registers_memory_limit": 1073741824,
      "max_register_size": 104857600,
      "max_number_registers": 100,
      "max_number_logs": 100,
      "max_total_log_length": 16384,
      "max_total_prepaid_gas": 300000000000000,
      "max_actions_per_receipt": 100,
      "max_number_bytes_method_names": 2000,
      "max_length_method_name": 256,
      "max_arguments_length": 4194304,
      "max_length_returned_data": 4194304,
      "max_contract_size": 4194304,
      "max_transaction_size": 4194304,
      "max_length_storage_key": 4194304,
      "max_length_storage_value": 4194304,
      "max_promises_per_function_call_action": 1024,
      "max_number_input_data_dependencies": 128,
      "max_functions_number_per_contract": 10000,
      "wasmer2_stack_limit": 102400,
      "account_id_validity_rules_version": 0
    }
  },
  "account_creation_config": {
    "min_allowed_top_level_account_length": 32,
    "registrar_account_id": "registrar"
  }
}
